apps/framework-cli/src/infrastructure/olap/clickhouse/inserter.rs

```diff
@@ -1,76 +1,308 @@
+        flush_interval: u64,
+        table: &str,
+        columns: Vec<String>,
+        commit_callback: OffsetCommitCallback,
+    ) -> Self {
+        let queue = Arc::new(Mutex::new(VecDeque::from([Batch::default()])));
+
+        let inserter = Self {
+            queue: queue.clone(),
+            client,
+            batch_size,
+            flush_interval,
+        };
 
-        tokio::spawn(flush(
-            clickhouse_config,
-            buffer.clone(),
-            table.to_string(),
-            columns,
-        ));
+        let inserter_clone = inserter.clone_for_flush();
+        let table_clone = table.to_string();
+        tokio::spawn(async move {
+            inserter_clone
+                .flush(table_clone, columns, commit_callback)
+                .await;
+        });
 
-        Self { buffer }
+        inserter
+    }
+
+    fn clone_for_flush(&self) -> Self {
+        Self {
+            queue: self.queue.clone(),
+            client: self.client.clone(),
+            batch_size: self.batch_size,
+            flush_interval: self.flush_interval,
+        }
     }
 
-    pub async fn insert(&self, record: ClickHouseRecord) -> anyhow::Result<()> {
-        let mut buffer: tokio::sync::MutexGuard<'_, Vec<ClickHouseRecord>> =
-            self.buffer.lock().await;
-        buffer.push(record);
+    pub async fn insert(
+        &self,
+        record: ClickHouseRecord,
+        partition: i32,
+        offset: i64,
+    ) -> anyhow::Result<()> {
+        let mut queue = self.queue.lock().await;
+
+        let current_batch = queue.back_mut().unwrap();
+
+        if current_batch.records.len() >= self.batch_size {
+            queue.push_back(Batch::default());
+            let new_batch = queue.back_mut().unwrap();
+            new_batch.records.push(record);
+            new_batch.update_offset(partition, offset);
+        } else {
+            current_batch.records.push(record);
+            current_batch.update_offset(partition, offset);
+        }
+
         Ok(())
     }
+
+    async fn flush(
+        &self,
+        table: String,
+        columns: Vec<String>,
+        commit_callback: OffsetCommitCallback,
+    ) {
+        let mut interval = time::interval(Duration::from_secs(self.flush_interval));
+
+        loop {
+            interval.tick().await;
+            let mut queue = self.queue.lock().await;
+
+            if queue.is_empty() || queue.front().map_or(true, |batch| batch.records.is_empty()) {
+                continue;
+            }
+
+            if let Some(batch) = queue.front() {
+                let batch_size = batch.records.len();
+
+                match self.client.insert(&table, &columns, &batch.records).await {
+                    Ok(_) => {
+                        debug!("Inserted {} records", batch_size,);
+
+                        for (partition, offset) in &batch.partition_offsets {
+                            let _ = commit_callback(*partition, *offset).map_err(|err| {
```

Comment:
i'm surprised to see that this is not an async function.

BTW `if let Err(e) =` looks nicer than `map_err` and discarding the result.

Reply:
The `store` function in the kafka lib is synchronous.
Changed the error pattern matching